Reduce MKL overheads on small shapes by not rewriting node to use MKL

milpuz01 opened the pull request:
We observed that when you run networks that have layers with smaller input and overall lower compute density overhead into MKL path dominates useful compute so in that case it is better to call default, Eigen, path.
 
This patch introduces changes to MKL layout pass where a liner analytical model is used to decide based on compute density and number of threads to either rewrite node to use MKL code path or leave it as it is. It propagates input shapes of the operation by decorating node with _input_shapes attribute number of threads used to parallelise operation by reading it from device on which operation is expected to execute (in this case CPUDevice) which, are then retrived in MKL layout pass and passed to the analytical model.
 
With this approach we have observed speedup for models available on TF-Hub as shown in the two attached graphs when running with 8 and 16 cores on Neoverse-V1 platform. Before shows results without this patch and after shows results with the patch relative when same models are ran with TF_ENABLE_ONEDNN_OPTS=0. Overall for 8 cores we see ~1.2x average speedup over TF_ENABLE_ONEDNN_OPTS=0 for 8 cores and ~1.12x for 16 cores.

![8cores_results](https://user-images.githubusercontent.com/79916358/225777461-0962b80c-4923-4b50-bb55-b65874ee3da9.png)

![16cores_results](https://user-images.githubusercontent.com/79916358/225777470-3fea315f-d78c-447b-9bbc-4f09202400da.png)

In addition to analytical model improvements for convolution, fused convolution, fused batch normalisation and swish, this patch also moves code for merging sigmoid and multiplication from remapper.cc to mkl_layout_pass.cc as it is MKL merge operation.

cc: @penpornk @cantonios

milpuz01 committed:
Reduce MKL overheads on small shapes by not rewriting node to use MKL

cantonios commented on tensorflow/core/common_runtime/mkl_layout_pass.cc:
```diff
@@ -968,6 +1004,11 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
  private:
   /// Maintain info about nodes to rewrite
   std::vector<RewriteInfo> rinfo_;
+  /// Mantain info about nodes to rewrite with additional
```
nit: spelling `Maintain`, and all comments are supposed to be sentences with punctuation (i.e. trailing period).

milpuz01 replied:
Thanks for the comment. Fixed in [59cc901](https://github.com/tensorflow/tensorflow/commit/59cc9014114ac5416e8980759ac6d105b9ae116d) and also all comments now have trailing period.

gzmkl commented on tensorflow/core/common_runtime/mkl_layout_pass.cc:
```diff
@@ -49,6 +50,18 @@ limitations under the License.
 namespace tensorflow {
 
+/// This table contains for each node name descriptors on which
+/// hardware  to check whether we should rewrite the operations
```
remove extra space between "hardware" and "to"

milpuz01 replied:
Thanks for spotting it. Fixed in [59cc901](https://github.com/tensorflow/tensorflow/commit/59cc9014114ac5416e8980759ac6d105b9ae116d)

milpuz01 committed:
Address comments from reviewers

milpuz01 commented on tensorflow/core/common_runtime/optimize_function_graph_utils.cc:
```diff
@@ -451,6 +451,13 @@ StatusOr<OptimizedFunctionGraphInfo> OptimizeFunctionGraph(
 
+  if (cpu_device->tensorflow_cpu_worker_threads() != nullptr) {
+    // Pass to the optimisation pass number of intra threads that are used to
+    // parallelise operations
```
Thanks for the comment. This is fixed now in [59cc901](https://github.com/tensorflow/tensorflow/commit/59cc9014114ac5416e8980759ac6d105b9ae116d).

milpuz01 commented on tensorflow/core/common_runtime/mkl_layout_pass.cc:
```diff
@@ -3032,6 +3182,129 @@ Node* MklLayoutRewritePass::CheckForNodeMerge(const Node* a) const {
 
+Status MklLayoutRewritePass::MergeSigmoidWithMul(std::unique_ptr<Graph>* g,
+                                                 Node* m, Node* n) {
+  CHECK_EQ(
```
Thanks. I have removed use of `CHECK` and `DCHECK`.

gbaned commented:
Hi @milpuz01 Can you please resolve conflicts? Thank you!

milpuz01 commented on tensorflow/core/common_runtime/mkl_layout_pass.cc:
```diff
@@ -49,6 +50,18 @@ limitations under the License.
 
+/// This table contains for each node name descriptors on which
+/// hardware  to check whether we should rewrite the operations
+/// to use MKL based on the parameters for heuristic
```
Thanks. I have removed occurrences where I used MKL to use oneDNN in [59cc901](https://github.com/tensorflow/tensorflow/commit/59cc9014114ac5416e8980759ac6d105b9ae116d)

penpornk commented on tensorflow/core/common_runtime/mkl_layout_pass.h:
```diff
@@ -25,6 +25,26 @@ limitations under the License.
+  // The model that is used to decide whether it is worth
+  // accelerating operations using oneDNN is:
+  //
+  // threshold = thread_synchronisation*thread_num + framework_tax
```
Nit:
```suggestion
  // threshold = thread_synchronisation * thread_num + framework_tax
```

penpornk reviewed (commented):
@gzmkl @TensorFlow-MKL Have you tested if this PR affects your performance?

penpornk reviewed (changes requested):
Some more comments. I'll post more tomorrow. Sorry for the delay!

milpuz01 committed:
Address comments from review

milpuz01 commented:
> Some more comments. I'll post more tomorrow. Sorry for the delay!

Thanks @penpornk. I have tried to address all your comments in https://github.com/tensorflow/tensorflow/pull/60026/commits/450f4feac896baa6b7d75d82b02973f73cf990b6

milpuz01 commented on tensorflow/core/common_runtime/mkl_layout_pass.cc:
```diff
@@ -49,6 +50,18 @@ limitations under the License.
+#ifdef DNNL_AARCH64_USE_ACL
+    {"Conv2D", 0x41, 0xd40, {0.9349, 22.603}},
+    {"_FusedConv2D", 0x41, 0xd40, {0.9349, 22.603}},
+    {"FusedBatchNormV3", 0x41, 0xd40, {0.3223, -0.8822}},
```
In our analysis we are comparing between oneDNN and Eigen code path and base cost is Eigen. There are few data points that we collect so for example for oneDNN+ACL we collect what is execution of ACL kernel, execution of oneDNN overhead when executing ACL kernel and overhead of calling oneDNN from TF. We do the same for Eigen such as time to call Eigen from TF and time to execute Eigen kernel like contraction across different number of shapes and different number of threads.

We then build regression model described above (from analysis we have tried to map/describe intercept value in the model as framework cost based on that that we collected) with base as Eigen so when you see negative value in that model that is effectively saying that with the metrics that we collected calling oneDNN from TF for that given operation is actually cheaper then doing it with Eigen so we account for that.

penpornk commented:
@TensorFlow-MKL Any feedback on this PR? Thank you!

milpuz01 commented:
> @TensorFlow-MKL Any feedback on this PR? Thank you!

@agramesh1, @mahmoud-abuzaina, @ashiqimranintel, @sachinmuradi  do you maybe have any comments about the PR? Many thanks.

gaurides commented:
Hi, we have couple main concerns :
1. Moving the fusion to mkl_layout_pass is not a good idea. From design perspective, we should preferably do all fusions in remapper with proper #ifdef INTEL_MKL and  IsMKLEnabled() checks. 
2. oneDNN would be the right place to handle optimizations for processor (or instruction set) variations instead of adding in mkl_layout_pass.

Both these changes applied for other fusions and primitives, can easily explode mkl_layout_pass code.

milpuz01 commented:
Hi @gaurides, thank you very much for your reply and the concerns that you have raised. I have couple of questions/comments that I have put below:

> 1. Moving the fusion to mkl_layout_pass is not a good idea. From design perspective, we should preferably do all fusions in remapper with proper #ifdef INTEL_MKL and  IsMKLEnabled() checks.

I can see that there are few other fusions already present in `mkl_layout_pass` such as merging convolution with bias (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/mkl_layout_pass.cc#L3048). Is there expectations that those fusions will also move to `remapper`?

> Both these changes applied for other fusions and primitives, can easily explode mkl_layout_pass code.

I think the same argument could be applied to `remapper` too as adding fusions there it will also make it explode in terms of code size? What influences decisions where graph rewriting pass should be? My impression was that lowering op to oneDNN should be in `mkl_layout_pass`, while fusion and other (high-level?) optimisations should be in `remapper` that might benefit other low-level libraries (such as Eigen).

> 2. oneDNN would be the right place to handle optimizations for processor (or instruction set) variations instead of adding in mkl_layout_pass.

The optimisation that we are doing with heuristic here is that we are choosing between lowering operation to be executed using oneDNN (i.e. rewriting it to be _Mkl) or executing operation using Eigen library so that is not possible to do in oneDNN as that decision is made before a call is made to oneDNN.

@penpornk, do you think that it might be introducing a new layout pass for AArch64 that would be called instead of `mkl_layout_pass`. We are already seeing some divergence in terms of fusion or lowering ops to onedNN such as here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/remapper.cc#L4603 and https://github.com/tensorflow/tensorflow/pull/60723 

Thanks,
Milos.

gaurides commented:
Hi @milpuz01, right, you will see some fusions currently happening in mkl_layout_pass but they are very old (5-6 years old). They need to be moved to remapper whenever possible. All new fusions have been added to remapper. We also added a pattern_matcher in remapper, which is easy and convenient to define patterns. take a [look here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/remapper.cc#L1444) . It's easier to match patterns than with code. So, you can try using it to do all fusions in grappler.
Using IsMKLEnabled() and #ifdef in remapper seems ok. You can probably define another functions that does the conditions checks to fuse or not. So, I recommend keeping fusions in remapper.
With respect to the logic to decide whether to use Mkl* version of the op or use Eigen, it's ok if you want to keep it in mkl_layout_pass as long as it doesn't affect Intel CPU code. But based on multiple HWs arch, it is easy to see it get complicated.  We typically try to use generic conditions for shape/size.

mdfaijul commented on tensorflow/core/common_runtime/mkl_layout_pass.cc:
```diff
@@ -4211,6 +4513,10 @@ Status MklLayoutRewritePass::Run(const GraphOptimizationPassOptions& options) {
+    num_intra_threads_ =
+        options.session_options->config.intra_op_parallelism_threads();
+  }
   auto process_graph = [&](std::unique_ptr<Graph>* g) {
```
Is there a default value set to `num_intra_threads_` when `options.session_options` is nullptr?

milpuz01 replied:
Hi @mdfaijul, in the implementation of `MklLayoutRewritePass` we initialise `num_intra_threads_` to the value of `port::MaxParallelism()` which gives number of schedulable CPUs which should be default number that is used when intra operation parallelism options is not set manually by user.

mdfaijul commented on tensorflow/core/grappler/optimizers/remapper.cc:
```diff
@@ -705,6 +705,23 @@ bool IsBiasSemanticAdd(const RemapperContext& ctx,
   return false;
 }
 
+void AddInputShapesAttr(const RemapperContext& ctx, int node_index) {
```
What if some dimensions are unknown like -1, -2 etc.

milpuz01 replied:
Hi @mdfaijul, in order to infer dimensions of the operation we are using method `InferStatically` from `graph_properties`. If it cannot infer dimensions or there are unknown then method `GetInputProperties()` would return empty array and we would not add `_input_shapes` attribute to nodes and as a result we would revert to default behaviour and not use dimensions to decide whether we want to rewrite not.

mdfaijul commented:
@milpuz01 Could you please add some unit tests.

gbaned commented:
Hi @milpuz01 Any update on this PR? Please. Thank you!

milpuz01 committed:
Address comments from reviewers

milpuz01 commented:
>  So, you can try using it to do all fusions in grappler. Using IsMKLEnabled() and #ifdef in remapper seems ok. You can probably define another functions that does the conditions checks to fuse or not. So, I recommend keeping fusions in remapper.

Thanks @gaurides for in-depth explanation. I have now reverted the change and put back fusion back that was removed from the mapper. In order to be able to decide whether to fuse or not on AArch64 I have added new function as per your suggestion in a new header file `mkl_heuristics.h` that can check now whether we want to fuse or not and use it in remapper. The same function is also used in layout pass too when decision needs to be made there. 

> With respect to the logic to decide whether to use Mkl* version of the op or use Eigen, it's ok if you want to keep it in mkl_layout_pass as long as it doesn't affect Intel CPU code. 

All the code that we are added is either guarded by `DNNL_AARCG64_USE_ACL` or reverts to default behaviour so it doesn't affect Intel's CPU code.

milpuz01 commented:
> @milpuz01 Could you please add some unit tests.

Thanks @mdfaijul for the comment. I have added new tests in `mkl_herustics_test.cc` to test method for calculating FLOPS and threshold. The rest of changes do not introduce any new behaviour that is not covered already by existing tests that are in tests for `remapper_test.cc` and `mkl_remapper_test.cc`. Furthermore, in this change PR: https://github.com/tensorflow/tensorflow/pull/60160 we have extended convolution benchmark to take into account rewriting of nodes using layout pass so that tests that we get best performance when running convolutions.

milpuz01 committed:
Remove definition of RewriteThreshold from mkl_layout_pass.h

milpuz01 committed:
Additional shapes for which to infer dimensions

milpuz01 committed:
Fix HWCAP_CPUID when not defined in some cases

cantonios commented on tensorflow/tsl/platform/cpu_info.cc:
```diff
@@ -345,6 +352,89 @@ void InitCPUIDInfo() {
+    }
+
+    std::ifstream CPUspresent;
+    CPUspresent.open("/sys/devices/system/cpu/present", std::ios::in);
```
Is this always going to work?  This seems like it would be very platform-dependent.

milpuz01 replied:
Thanks @cantonios for the comment. Yes, as you have said it is platform dependent. I have made updates in 528c1dc to address this. For platforms that pass through `ifdefs` if file doesn't exists we will simply just revert to default behaviour as will not be picking up heuristics.

milpuz01 committed:
Address comments from review

cantonios commented on tensorflow/core/common_runtime/mkl_layout_pass.cc:
```diff
@@ -425,9 +429,12 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    rinfothr_.push_back(
+        {{csinfo_.conv2d, mkl_op_registry::GetMklOpName(csinfo_.conv2d),
+          CopyAttrsConvCheckConstFilter,
+          std::function<bool(const Node*)>(),  // we set this function to empty
```
nit: comments should start with capital, end with punctuation.

milpuz01 replied:
Thanks @cantonios. Fixed in 5cfd215.

milpuz01 committed:
Change to comment to start with capital letter and end with fullstop

milpuz01 commented on tensorflow/core/common_runtime/mkl_layout_pass.cc:
```diff
@@ -1769,7 +1810,25 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    double total_mflops = CalculateNodeMFlops(n->attrs(), n->type_string());
+    double thr = FindRewriteThreshold(n->type_string(), threads);
+
+    return true ? (total_mflops < 0 || total_mflops >= thr) : false;
```
Thanks @penpornk. Fixed in cdc2fdd.

penpornk reviewed (approved):
Thank you!

penpornk commented on tensorflow/core/util/mkl_heuristics.h:
```diff
@@ -0,0 +1,123 @@
+namespace tensorflow {
+
+struct RewriteThreshold {
+  string op;
```
```suggestion
  std::string op;
```

milpuz01 replied:
Fixed in [3abb8b7](https://github.com/tensorflow/tensorflow/pull/60026/commits/3abb8b750a628d5a115737a5b335f519ec61f868)

penpornk commented on tensorflow/core/util/mkl_heuristics.h:
```diff
@@ -0,0 +1,123 @@
+#define TENSORFLOW_CORE_UTIL_MKL_HEURISTICS_H_
+#ifdef INTEL_MKL
+
+#include "tensorflow/tsl/platform/cpu_info.h"
```
Please add `#include <vector>` above this line. Please also add a blank line to separate from `#include "tensorflow/tsl/platform/cpu_info.h"`.

milpuz01 replied:
Fixed in [3abb8b7](https://github.com/tensorflow/tensorflow/pull/60026/commits/3abb8b750a628d5a115737a5b335f519ec61f868)

penpornk reviewed (changes requested):
Could you please help fix [PY+CPP Ubuntu CPU](https://source.cloud.google.com/results/invocations/c09a2a5f-e1e3-410f-8b8f-d2b79592ea1d/log) build failures?

```
In file included from tensorflow/core/util/mkl_heuristics_test.cc:19:
./tensorflow/core/util/mkl_heuristics.h:28:3: error: unknown type name 'string'; did you mean 'std::string'?
  string op;
  ^~~~~~
  std::string
...
In file included from tensorflow/core/util/mkl_heuristics_test.cc:19:
./tensorflow/core/util/mkl_heuristics.h:82:8: error: no member named 'vector' in namespace 'std'
  std::vector<const TensorShapeProto*> shape_attrs;
  ~~~~~^
...
```

milpuz01 committed:
Correct ordering to include headers for heuristics test

milpuz01 commented:
> Could you please help fix [PY+CPP Ubuntu CPU](https://source.cloud.google.com/results/invocations/c09a2a5f-e1e3-410f-8b8f-d2b79592ea1d/log) build failures?

Sorry about this @penporn. I believe that commit [3abb8b7](https://github.com/tensorflow/tensorflow/pull/60026/commits/3abb8b750a628d5a115737a5b335f519ec61f868) should fix the failures as ordering of included headers was not correct in `mkl_heuristics_test.cc`

milpuz01 committed:
Fix failure to build Mkl heuristics test

penpornk reviewed (approved):
Thank you for the changes!

penpornk commented:
Could you please help take a look at `//tensorflow/core/common_runtime/eager:context_test` failure?
```
==================== Test output for //tensorflow/core/common_runtime/eager:context_test:
2023-07-25 12:21:47.692537: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2023-07-25 12:21:47.692765: W tensorflow/tsl/lib/monitoring/collection_registry.cc:81] Trying to register 2 metrics with the same name: /tensorflow/core/bfc_allocator_delay. The old value will be erased in order to register a new one. Please check if you link the metric more than once, or if the name is already used by other metrics.
[==========] Running 10 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 10 tests from EagerContextTest
[ RUN      ] EagerContextTest.CompositeDevice
[       OK ] EagerContextTest.CompositeDevice (25 ms)
[ RUN      ] EagerContextTest.CompositeDeviceWithGivenName
[       OK ] EagerContextTest.CompositeDeviceWithGivenName (0 ms)
[ RUN      ] EagerContextTest.AddFunctionDef
[       OK ] EagerContextTest.AddFunctionDef (26 ms)
[ RUN      ] EagerContextTest.AddFunctionDefRepeatSame
[       OK ] EagerContextTest.AddFunctionDefRepeatSame (0 ms)
[ RUN      ] EagerContextTest.AddFunctionDefRepeatDifferent
[       OK ] EagerContextTest.AddFunctionDefRepeatDifferent (1 ms)
[ RUN      ] EagerContextTest.FunctionErrorRecovery
2023-07-25 12:21:47.827207: F ./tensorflow/core/framework/device_base.h:146] Check failed: cpu_worker_threads_ != nullptr 
*** Received signal 6 ***
*** BEGIN MANGLED STACK TRACE ***
```
I believe this comes from the changes in //tensorflow/core/common_runtime/optimize_function_graph_utils.cc.
https://github.com/tensorflow/tensorflow/pull/60026/files#diff-cf5de83fc2a1e4038beb6f3994daae6bb9bc321f8eb05a6e125980ab4b598680R556-R562

milpuz01 committed:
Fixes for failing tests from CI

milpuz01 commented:
> Could you please help take a look at `//tensorflow/core/common_runtime/eager:context_test` failure?

Thanks @penpornk. Commit 2701cc4 should make the test to pass. I have restricted to only check for number of threads for intra op if it is local device because that is common path that is used when doing MKL layout pass. I have also moved to set it inside of `PreprocessAndPartitionGraph()` instead of `OptimizeFunctionGraph()` as MKL layout pass is called from former.

penpornk reviewed (approved):
Thank you for the fixes!

penpornk commented on tensorflow/tsl/platform/cpu_info.cc:
```diff
@@ -22,6 +22,13 @@ limitations under the License.
 #include <mutex>  // NOLINT
 #endif
+#if defined(PLATFORM_IS_ARM64)
+#include <sys/auxv.h>
```
Is this header needed? What function is it needed for? (I don't think `getauxval` needs this header.) 
It broke MacOS build. If it's not needed, I'll just remove it. If it is. I'll add a guard for Apple.

milpuz01 replied:
HI @penpornk, in that header are HWCAP bits defined that we are using to check for presence of CPUID registers. I would suggest to add guard for Apple and also OpenBSD in case TF is build there too.

`#if !defined(__APPLE__) && !defined(__OpenBSD__) && defined(PLATFORM_IS_ARM64)`

Sorry about breakage.

penpornk replied:
Should I just change everything about cpu info (CPUIDInfo class, etc) that has `#ifdef PLATFORM_IS_ARM64` to `#if defined(PLATFORM_IS_ARM64) && !defined(__APPLE__) && !defined(__OpenBSD__)` ? Or just guarding the header should be enough (i.e., other code should be generic enough for other platforms)?

milpuz01 replied:
We have put the check `#if !defined(PLATFORM_WINDOWS) && !defined(__APPLE__) && !defined(__OpenBSD__)`  around reading files (see here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tsl/platform/cpu_info.cc#L380), but on the second thoughts I think it might be better to put as you have suggested around all instances of `PLATFORM_IS_ARM64` in `cpu_info.cc` because in `mkl_heuristics.h` we are using `tsl::port::CPUFamily()` and `tsl::port::CPUModelName()` and that will return 0 on platforms that do not confirm to `ifdef`.

Thanks!
